### PR TITLE
[POAE7-497] PMem extension implementation for UnsafeExternalSorter

### PR DESCRIPTION
**Pull request by yma11**
### Why are the changes needed?
1. memory spill to PMem for UnsafeExternalSorter
2. add memory manager for PMem

### Does this PR introduce _any_ user-facing change?
Yes

### How was this patch tested?
UT

### CODE REVIEW

**Comment by yma11** [comment]
@winningsix @offthewall123 Please take a look. Thanks.

File: core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala [hunk 1]
```diff
@@ -148,6 +148,15 @@ private[spark] class UnifiedMemoryManager(
       numBytes, taskAttemptId, maybeGrowExecutionPool, () => computeMaxExecutionPoolSize)
   }
 
+  override def acquireExtendedMemory(numBytes: Long, taskAttemptId: Long): Long = synchronized {
```
Comment (offthewall123) [inline core/src/main/scala/org/apache/spark/memory/UnifiedMemoryManager.scala:151]: Minor comment here, as for the name of the method, maybe return Boolean is more suitable?
I think it's strange to return the param `numBytes` back, any consideration here?

File: core/src/main/scala/org/apache/spark/memory/ExtendedMemoryPool.scala [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    * Release all memory for the given task and mark it as inactive (e.g. when a task ends).
+    *
+    * @return the number of bytes freed.
+    */
+  def releaseAllMemoryForTask(taskAttemptId: Long): Long = lock.synchronized {
+    val numBytesToFree = getMemoryUsageForTask(taskAttemptId)
+    releaseMemory(numBytesToFree, taskAttemptId)
+    numBytesToFree
+  }
+}
```
Comment (offthewall123) [inline core/src/main/scala/org/apache/spark/memory/ExtendedMemoryPool.scala:106]: Add newLine at the end of file

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java [hunk 3]
```diff
@@ -440,7 +478,6 @@ public void insertKVRecord(Object keyBase, long keyOffset, int keyLen,
     pageCursor += keyLen;
     Platform.copyMemory(valueBase, valueOffset, base, pageCursor, valueLen);
     pageCursor += valueLen;
-
```
Comment (offthewall123) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java:443]: Please revert it if it's the origin spark empty line

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemReader.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        if (recordLength > arr.length) {
+            arr = new byte[recordLength];
+            baseObject = arr;
+        }
+        Platform.copyMemory(null, offset , baseObject, Platform.BYTE_ARRAY_OFFSET, recordLength);
+        offset += recordLength;
+        readedRecordsInCurrentPage ++;
+        numRecordsRemaining --;
+
+
```
Comment (offthewall123) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemReader.java:51]: Remove unnecessary empty line.

File: core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java [hunk 5]
```diff
@@ -350,6 +364,66 @@ public void freePage(MemoryBlock page, MemoryConsumer consumer) {
...
+        throw new IllegalStateException(
+                "Have already allocated a maximum of " + PAGE_TABLE_SIZE + " pages");
+      }
+      allocatedPages.set(pageNumber);
+    }
+    MemoryBlock page = null;
+    try {
+      page = memoryManager.extendedMemoryAllocator().allocate(size);
+    } catch (OutOfMemoryError e) {
+      logger.error("Failed to allocate a PMem page ({} bytes).", size);
```
Comment (offthewall123) [inline core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java:391]: if OutOfMemory,  dircetly return here?

File: common/unsafe/src/main/java/org/apache/spark/unsafe/memory/ExtendedMemoryAllocator.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        return memoryBlock;
+    }
+
+    @Override
+    public void free(MemoryBlock memoryBlock) {
+        assert (memoryBlock.getBaseObject() == null) :
+                "baseObject not null; are you trying to use the AEP-heap allocator to free on-heap memory?";
+        PersistentMemoryPlatform.freeMemory(memoryBlock.getBaseOffset());
+    }
+}
```
Comment (offthewall123) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/memory/ExtendedMemoryAllocator.java:21]: Add an empty line at the end of file.

File: core/src/main/scala/org/apache/spark/SparkEnv.scala [hunk 7]
```diff
@@ -25,14 +25,13 @@ import scala.collection.JavaConverters._
 import scala.collection.concurrent
 import scala.collection.mutable
 import scala.util.Properties
-
 import com.google.common.cache.CacheBuilder
+import com.intel.oap.common.unsafe.PersistentMemoryPlatform
 import org.apache.hadoop.conf.Configuration
-
```
Comment (offthewall123) [inline core/src/main/scala/org/apache/spark/SparkEnv.scala:31]: can you pass checkstyle if delete these empty line?

File: core/src/main/scala/org/apache/spark/SparkEnv.scala [hunk 8]
```diff
@@ -25,14 +25,13 @@ import scala.collection.JavaConverters._
 import scala.collection.concurrent
 import scala.collection.mutable
 import scala.util.Properties
-
```
Comment (offthewall123) [inline core/src/main/scala/org/apache/spark/SparkEnv.scala:28]: same here

File: core/pom.xml [hunk 9]
```diff
@@ -35,6 +35,11 @@
   </properties>
   
   <dependencies>
+    <dependency>
+      <groupId>com.intel</groupId>
+      <artifactId>oap</artifactId>
```
Comment (offthewall123) [inline core/pom.xml:40]: to import oap-common 0.9 it is 
<dependency>
      <groupId>com.intel.oap</groupId>
      <artifactId>oap-common</artifactId>
      <version>0.9.0</version>
    </dependency>

Comment (offthewall123) [reply core/pom.xml:40]: Did you build a private oap-common jar locally?

File: common/unsafe/pom.xml [hunk 10]
```diff
@@ -35,6 +35,11 @@
   </properties>
 
   <dependencies>
+    <dependency>
+      <groupId>com.intel</groupId>
```
Comment (jikunshang) [inline common/unsafe/pom.xml:39]: seems it's not same with current oap side. see  https://github.com/Intel-bigdata/OAP/blob/master/oap-common/pom.xml#L25

**Review by jikunshang — APPROVED** [review]
Overall LGTM. Will merge it for weekend test, Please address comments in next PR.

**Pull request merged**